File: server/geometry.go
package main

// Geometrie des Grundrisses -- die Go-Entsprechung von
// client/src/model/geometry.js.
//
// Der Server braucht sie fuer zwei Dinge: die Kennzahlen der Projektuebersicht
// (Raumzahl und Flaeche) und den serverseitigen SVG-Export. Beide Seiten
// muessen dieselben Raeume finden, deshalb pruefen beide dieselben Faelle --
// siehe geometry_test.go und tests/geometry.test.js.
//
// Alle Laengen sind Zentimeter, Winkel im Bogenmass.

import (
	"math"
	"sort"
)

const eps = 1e-6

type Point struct {
	X float64 `json:"x"`
	Y float64 `json:"y"`
}

type Wall struct {
	ID          string  `json:"id"`
	A           Point   `json:"a"`
	B           Point   `json:"b"`
	ThicknessCm float64 `json:"thicknessCm"`
	HeightCm    float64 `json:"heightCm"`
}

type Opening struct {
	ID       string  `json:"id"`
	WallID   string  `json:"wallId"`
	OffsetCm float64 `json:"offsetCm"`
	WidthCm  float64 `json:"widthCm"`
	HeightCm float64 `json:"heightCm"`
	SillCm   float64 `json:"sillCm"`
	Type     string  `json:"type"`
	Swing    int     `json:"swing"`
}

type Furniture struct {
	ID          string  `json:"id"`
	CatalogID   string  `json:"catalogId"`
	Label       string  `json:"label"`
	X           float64 `json:"x"`
	Y           float64 `json:"y"`
	WidthCm     float64 `json:"widthCm"`
	DepthCm     float64 `json:"depthCm"`
	HeightCm    float64 `json:"heightCm"`
	RotationDeg float64 `json:"rotationDeg"`
	Color       string  `json:"color"`
}

type Level struct {
	ID        string      `json:"id"`
	Name      string      `json:"name"`
	HeightCm  float64     `json:"heightCm"`
	Walls     []Wall      `json:"walls"`
	Openings  []Opening   `json:"openings"`
	Furniture []Furniture `json:"furniture"`
}

type Project struct {
	Version int     `json:"version"`
	Name    string  `json:"name"`
	GridCm  float64 `json:"gridCm"`
	Levels  []Level `json:"levels"`
}

type Room struct {
	Points []Point `json:"points"`
	Area   float64 `json:"area"`
}

type segment struct{ A, B Point }

func dist(a, b Point) float64 { return math.Hypot(b.X-a.X, b.Y-a.Y) }

// closestOnSegment liefert den Parameter t in [0,1] und den Abstand.
func closestOnSegment(p, a, b Point) (t, d float64) {
	abx, aby := b.X-a.X, b.Y-a.Y
	l2 := abx*abx + aby*aby
	if l2 < eps {
		return 0, dist(p, a)
	}
	t = ((p.X-a.X)*abx + (p.Y-a.Y)*aby) / l2
	t = math.Max(0, math.Min(1, t))
	q := Point{a.X + abx*t, a.Y + aby*t}
	return t, dist(p, q)
}

func segmentIntersection(a1, a2, b1, b2 Point) (Point, bool) {
	rx, ry := a2.X-a1.X, a2.Y-a1.Y
	sx, sy := b2.X-b1.X, b2.Y-b1.Y
	den := rx*sy - ry*sx
	if math.Abs(den) < eps {
		return Point{}, false
	}
	t := ((b1.X-a1.X)*sy - (b1.Y-a1.Y)*sx) / den
	u := ((b1.X-a1.X)*ry - (b1.Y-a1.Y)*rx) / den
	if t < -eps || t > 1+eps || u < -eps || u > 1+eps {
		return Point{}, false
	}
	return Point{a1.X + rx*t, a1.Y + ry*t}, true
}

// PolygonArea ist positiv bei mathematisch positivem Umlaufsinn.
func PolygonArea(pts []Point) float64 {
	a := 0.0
	for i := range pts {
		p, q := pts[i], pts[(i+1)%len(pts)]
		a += p.X*q.Y - q.X*p.Y
	}
	return a / 2
}

func PolygonCentroid(pts []Point) Point {
	a := PolygonArea(pts)
	if math.Abs(a) < eps {
		var sx, sy float64
		for _, p := range pts {
			sx, sy = sx+p.X, sy+p.Y
		}
		n := math.Max(float64(len(pts)), 1)
		return Point{sx / n, sy / n}
	}
	var cx, cy float64
	for i := range pts {
		p, q := pts[i], pts[(i+1)%len(pts)]
		f := p.X*q.Y - q.X*p.Y
		cx += (p.X + q.X) * f
		cy += (p.Y + q.Y) * f
	}
	return Point{cx / (6 * a), cy / (6 * a)}
}

// nodeSegments zerlegt Strecken an allen Beruehrpunkten.
//
// Ohne diesen Schritt haengt eine Zwischenwand, die mitten auf einer
// durchgehenden Aussenwand endet, im Graphen in der Luft: der T-Stoss waere
// kein Knoten, und die Flaechensuche saehe beide Raeume als einen.
func nodeSegments(segs []segment, tol float64) []segment {
	endpoints := make([]Point, 0, len(segs)*2)
	for _, s := range segs {
		endpoints = append(endpoints, s.A, s.B)
	}

	out := []segment{}
	for i, seg := range segs {
		l := dist(seg.A, seg.B)
		if l < tol {
			continue
		}
		cuts := []float64{}
		add := func(p Point) {
			t, d := closestOnSegment(p, seg.A, seg.B)
			along := t * l
			// Nur echte Zwischenpunkte -- die eigenen Enden teilen nichts.
			if d <= tol && along > tol && along < l-tol {
				cuts = append(cuts, along)
			}
		}
		for _, p := range endpoints {
			add(p)
		}
		for j, other := range segs {
			if i == j {
				continue
			}
			if x, ok := segmentIntersection(seg.A, seg.B, other.A, other.B); ok {
				add(x)
			}
		}
		sort.Float64s(cuts)

		dx, dy := (seg.B.X-seg.A.X)/l, (seg.B.Y-seg.A.Y)/l
		prev := 0.0
		for _, c := range cuts {
			if c-prev < tol {
				continue
			}
			out = append(out, segment{
				A: Point{seg.A.X + dx*prev, seg.A.Y + dy*prev},
				B: Point{seg.A.X + dx*c, seg.A.Y + dy*c},
			})
			prev = c
		}
		if l-prev > tol {
			out = append(out, segment{
				A: Point{seg.A.X + dx*prev, seg.A.Y + dy*prev},
				B: seg.B,
			})
		}
	}
	return out
}

type halfEdge struct {
	from, to int
	angle    float64
	visited  bool
}

// FindRooms findet die geschlossenen Raeume in einem Wandnetz.
//
// Die Wandmittellinien bilden einen planaren Graphen; jede innere Flaeche ist
// ein Raum. Ermittelt wird sie ueber Half-Edge-Traversal: an jedem Knoten wird
// immer die naechste Kante im Uhrzeigersinn genommen, dadurch laeuft man die
// minimalen Zyklen ab. Die Aussenflaeche faellt weg, weil sie als einzige den
// umgekehrten Umlaufsinn hat.
func FindRooms(walls []Wall, tol float64) []Room {
	segs := make([]segment, 0, len(walls))
	for _, w := range walls {
		segs = append(segs, segment{A: w.A, B: w.B})
	}
	noded := nodeSegments(segs, tol)

	nodes := []Point{}
	key := func(p Point) int {
		for i, n := range nodes {
			if dist(n, p) <= tol {
				return i
			}
		}
		nodes = append(nodes, p)
		return len(nodes) - 1
	}

	edges := []halfEdge{}
	outgoing := map[int][]int{}
	for _, s := range noded {
		from, to := key(s.A), key(s.B)
		if from == to {
			continue
		}
		for _, pair := range [2][2]int{{from, to}, {to, from}} {
			a, b := pair[0], pair[1]
			idx := len(edges)
			edges = append(edges, halfEdge{
				from:  a,
				to:    b,
				angle: math.Atan2(nodes[b].Y-nodes[a].Y, nodes[b].X-nodes[a].X),
			})
			outgoing[a] = append(outgoing[a], idx)
		}
	}
	for _, list := range outgoing {
		sort.Slice(list, func(i, j int) bool { return edges[list[i]].angle < edges[list[j]].angle })
	}

	twin := func(i int) int {
		if i%2 == 0 {
			return i + 1
		}
		return i - 1
	}

	next := func(idx int) int {
		he := edges[idx]
		list := outgoing[he.to]
		if len(list) == 0 {
			return -1
		}
		back := math.Atan2(nodes[he.from].Y-nodes[he.to].Y, nodes[he.from].X-nodes[he.to].X)
		best, bestDelta := -1, math.Inf(1)
		for _, cand := range list {
			delta := back - edges[cand].angle
			for delta <= eps {
				delta += 2 * math.Pi
			}
			if delta < bestDelta {
				bestDelta, best = delta, cand
			}
		}
		if best == -1 {
			return twin(idx) // Sackgasse: zurueck
		}
		return best
	}

	rooms := []Room{}
	for start := range edges {
		if edges[start].visited {
			continue
		}
		cycle := []int{}
		cur, guard := start, 0
		for !edges[cur].visited && guard < len(edges)+2 {
			guard++
			edges[cur].visited = true
			cycle = append(cycle, cur)
			cur = next(cur)
			if cur < 0 {
				break
			}
		}
		if len(cycle) < 3 {
			continue
		}
		pts := make([]Point, 0, len(cycle))
		for _, i := range cycle {
			pts = append(pts, nodes[edges[i].from])
		}
		area := PolygonArea(pts)
		// Nur Flaechen mit positivem Umlaufsinn sind Innenraeume; die
		// Aussenhuelle laeuft andersherum und wird verworfen.
		if area > 100 {
			rooms = append(rooms, Room{Points: pts, Area: area})
		}
	}
	return rooms
}

// WallSolid ist ein massives Stueck einer Wand zwischen den Oeffnungen.
type WallSolid struct {
	From, To    float64
	Bottom, Top float64
}

// WallSolids zerlegt eine Wand in ihre massiven Stuecke. Bruestung unter dem
// Fenster und Sturz darueber entstehen dabei von selbst -- ohne Boolesche
// Operationen auf Geometrie.
func WallSolids(w Wall, ops []Opening) []WallSolid {
	total := dist(w.A, w.B)
	height := w.HeightCm
	if height <= 0 {
		height = 250
	}

	type gap struct{ start, end, sill, top float64 }
	gaps := []gap{}
	for _, o := range ops {
		g := gap{
			start: math.Max(0, o.OffsetCm-o.WidthCm/2),
			end:   math.Min(total, o.OffsetCm+o.WidthCm/2),
			sill:  o.SillCm,
			top:   math.Min(height, o.SillCm+o.HeightCm),
		}
		if g.end > g.start {
			gaps = append(gaps, g)
		}
	}
	sort.Slice(gaps, func(i, j int) bool { return gaps[i].start < gaps[j].start })

	out := []WallSolid{}
	cursor := 0.0
	for _, g := range gaps {
		if g.start > cursor {
			out = append(out, WallSolid{cursor, g.start, 0, height})
		}
		if g.sill > 0 {
			out = append(out, WallSolid{g.start, g.end, 0, g.sill})
		}
		if g.top < height {
			out = append(out, WallSolid{g.start, g.end, g.top, height})
		}
		cursor = math.Max(cursor, g.end)
	}
	if cursor < total {
		out = append(out, WallSolid{cursor, total, 0, height})
	}
	return out
}

File: src/model/geometry.ts
// Reine Geometrie -- keine React-, Canvas- oder Store-Abhaengigkeiten,
// damit das hier ohne Browser testbar bleibt.

import type { Cm, Point, Rad, Room, Segment } from './types.ts';

export const EPS = 1e-6;

export const sub = (a: Point, b: Point): Point => ({ x: a.x - b.x, y: a.y - b.y });
export const add = (a: Point, b: Point): Point => ({ x: a.x + b.x, y: a.y + b.y });
export const scale = (a: Point, s: number): Point => ({ x: a.x * s, y: a.y * s });
export const dot = (a: Point, b: Point): number => a.x * b.x + a.y * b.y;
export const cross = (a: Point, b: Point): number => a.x * b.y - a.y * b.x;
export const len = (a: Point): number => Math.hypot(a.x, a.y);
export const dist = (a: Point, b: Point): number => Math.hypot(b.x - a.x, b.y - a.y);

export function normalize(a: Point): Point {
  const l = len(a);
  return l < EPS ? { x: 0, y: 0 } : { x: a.x / l, y: a.y / l };
}

/** Normale (90 Grad links) eines Vektors. */
export function normal(a: Point): Point {
  return { x: -a.y, y: a.x };
}

export function rotate(p: Point, rad: Rad, origin: Point = { x: 0, y: 0 }): Point {
  const c = Math.cos(rad);
  const s = Math.sin(rad);
  const d = sub(p, origin);
  return { x: origin.x + d.x * c - d.y * s, y: origin.y + d.x * s + d.y * c };
}

export function angleOf(v: Point): Rad {
  return Math.atan2(v.y, v.x);
}

/** Winkel auf ein Vielfaches von stepRad einrasten. */
export function snapAngle(a: Point, b: Point, stepRad: Rad): Point {
  const v = sub(b, a);
  const l = len(v);
  if (l < EPS) return b;
  const snapped = Math.round(angleOf(v) / stepRad) * stepRad;
  return { x: a.x + Math.cos(snapped) * l, y: a.y + Math.sin(snapped) * l };
}

/** Punkt auf einer Strecke, der p am naechsten liegt -- plus Parameter t in [0,1]. */
export function closestPointOnSegment(p: Point, a: Point, b: Point): { point: Point; t: number; dist: number } {
  const ab = sub(b, a);
  const l2 = dot(ab, ab);
  if (l2 < EPS) return { point: { ...a }, t: 0, dist: dist(p, a) };
  let t = dot(sub(p, a), ab) / l2;
  t = Math.max(0, Math.min(1, t));
  const point = add(a, scale(ab, t));
  return { point, t, dist: dist(p, point) };
}

export function pointSegmentDistance(p: Point, a: Point, b: Point): number {
  return closestPointOnSegment(p, a, b).dist;
}

/** Schnittpunkt zweier Strecken oder null. */
export function segmentIntersection(a1: Point, a2: Point, b1: Point, b2: Point): Point | null {
  const r = sub(a2, a1);
  const s = sub(b2, b1);
  const denom = cross(r, s);
  if (Math.abs(denom) < EPS) return null;
  const t = cross(sub(b1, a1), s) / denom;
  const u = cross(sub(b1, a1), r) / denom;
  if (t < -EPS || t > 1 + EPS || u < -EPS || u > 1 + EPS) return null;
  return add(a1, scale(r, t));
}

/** Flaeche eines Polygons (positiv bei mathematisch positivem Umlaufsinn). */
export function polygonArea(points: Point[]): number {
  let a = 0;
  for (let i = 0; i < points.length; i++) {
    const p = points[i];
    const q = points[(i + 1) % points.length];
    a += p.x * q.y - q.x * p.y;
  }
  return a / 2;
}

export function polygonCentroid(points: Point[]): Point {
  const a = polygonArea(points);
  if (Math.abs(a) < EPS) {
    // Entartetes Polygon: einfacher Mittelwert statt Division durch ~0.
    const sum = points.reduce((acc, p) => add(acc, p), { x: 0, y: 0 });
    return scale(sum, 1 / Math.max(points.length, 1));
  }
  let cx = 0;
  let cy = 0;
  for (let i = 0; i < points.length; i++) {
    const p = points[i];
    const q = points[(i + 1) % points.length];
    const f = p.x * q.y - q.x * p.y;
    cx += (p.x + q.x) * f;
    cy += (p.y + q.y) * f;
  }
  return { x: cx / (6 * a), y: cy / (6 * a) };
}

export function pointInPolygon(p: Point, points: Point[]): boolean {
  let inside = false;
  for (let i = 0, j = points.length - 1; i < points.length; j = i++) {
    const pi = points[i];
    const pj = points[j];
    const intersects =
      pi.y > p.y !== pj.y > p.y &&
      p.x < ((pj.x - pi.x) * (p.y - pi.y)) / (pj.y - pi.y + EPS) + pi.x;
    if (intersects) inside = !inside;
  }
  return inside;
}

/** Achsparallele Huelle einer Punktmenge. */
export function bboxOf(points: Point[]): { minX: number; minY: number; maxX: number; maxY: number } {
  if (!points.length) return { minX: 0, minY: 0, maxX: 0, maxY: 0 };
  let minX = Infinity;
  let minY = Infinity;
  let maxX = -Infinity;
  let maxY = -Infinity;
  for (const p of points) {
    if (p.x < minX) minX = p.x;
    if (p.y < minY) minY = p.y;
    if (p.x > maxX) maxX = p.x;
    if (p.y > maxY) maxY = p.y;
  }
  return { minX, minY, maxX, maxY };
}

/** Die vier Eckpunkte eines gedrehten Rechtecks (Mittelpunkt, Breite, Tiefe, Winkel). */
export function rectCorners(cx: Cm, cy: Cm, w: Cm, d: Cm, rad: Rad): Point[] {
  const hw = w / 2;
  const hd = d / 2;
  const c = { x: cx, y: cy };
  return [
    rotate({ x: cx - hw, y: cy - hd }, rad, c),
    rotate({ x: cx + hw, y: cy - hd }, rad, c),
    rotate({ x: cx + hw, y: cy + hd }, rad, c),
    rotate({ x: cx - hw, y: cy + hd }, rad, c),
  ];
}

/** Liegt p im gedrehten Rechteck? Rechnet in dessen lokalem System. */
export function pointInRect(p: Point, cx: Cm, cy: Cm, w: Cm, d: Cm, rad: Rad): boolean {
  const local = rotate(p, -rad, { x: cx, y: cy });
  return (
    local.x >= cx - w / 2 &&
    local.x <= cx + w / 2 &&
    local.y >= cy - d / 2 &&
    local.y <= cy + d / 2
  );
}

/**
 * Zerlegt Strecken an allen Beruehrpunkten ("Noding").
 *
 * Ohne diesen Schritt haengt eine Zwischenwand, die mitten auf einer
 * durchgehenden Aussenwand endet, im Graphen in der Luft: der T-Stoss waere
 * kein Knoten, und die Flaechensuche wuerde beide Raeume als einen sehen.
 * Beruecksichtigt werden Endpunkte anderer Strecken auf der Strecke sowie
 * echte Kreuzungen.
 */
export function nodeSegments(segments: Segment[], tol = 1): Segment[] {
  const endpoints = [];
  for (const s of segments) {
    endpoints.push(s.a, s.b);
  }

  const result: Segment[] = [];
  for (let i = 0; i < segments.length; i++) {
    const seg = segments[i];
    const l = dist(seg.a, seg.b);
    if (l < tol) continue;

    const cuts: number[] = [];
    const addCut = (p: Point) => {
      const { t, dist: d } = closestPointOnSegment(p, seg.a, seg.b);
      const along = t * l;
      // Nur echte Zwischenpunkte -- die eigenen Enden teilen nichts.
      if (d <= tol && along > tol && along < l - tol) cuts.push(along);
    };

    for (const p of endpoints) addCut(p);
    for (let j = 0; j < segments.length; j++) {
      if (j === i) continue;
      const x = segmentIntersection(seg.a, seg.b, segments[j].a, segments[j].b);
      if (x) addCut(x);
    }

    cuts.sort((a, b) => a - b);
    const dir = scale(sub(seg.b, seg.a), 1 / l);
    let prev = 0;
    for (const cut of cuts) {
      if (cut - prev < tol) continue;
      result.push({ a: add(seg.a, scale(dir, prev)), b: add(seg.a, scale(dir, cut)) });
      prev = cut;
    }
    if (l - prev > tol) {
      result.push({ a: add(seg.a, scale(dir, prev)), b: { ...seg.b } });
    }
  }
  return result;
}

/**
 * Findet die geschlossenen Raeume in einem Wandnetz.
 *
 * Die Wandmittellinien bilden einen planaren Graphen; jede innere Flaeche
 * dieses Graphen ist ein Raum. Ermittelt wird sie ueber Half-Edge-Traversal:
 * an jedem Knoten wird immer die naechste Kante im Uhrzeigersinn genommen,
 * dadurch laeuft man die minimalen Zyklen ab. Die Aussenflaeche faellt weg,
 * weil sie als einzige den umgekehrten Umlaufsinn hat.
 *
 * @param {Array<{a:{x,y}, b:{x,y}}>} segments Wandmittellinien
 * @param {number} tol Knoten naeher als tol (cm) gelten als derselbe Punkt
 * @returns {Array<{points: Array<{x,y}>, area: number}>} Raumpolygone, cm bzw. cm2
 */
export function findRooms(segments: Segment[], tol = 1): Room[] {
  const noded = nodeSegments(segments, tol);
  const nodes: Point[] = [];
  const keyOf = (p: Point) => {
    for (let i = 0; i < nodes.length; i++) {
      if (dist(nodes[i], p) <= tol) return i;
    }
    nodes.push({ x: p.x, y: p.y });
    return nodes.length - 1;
  };

  const halfEdges: { from: number; to: number; angle: Rad; visited: boolean }[] = [];
  const outgoing = new Map<number, number[]>(); // Knoten -> Indizes der ausgehenden Half-Edges

  for (const seg of noded) {
    const from = keyOf(seg.a);
    const to = keyOf(seg.b);
    if (from === to) continue;
    for (const [s, e] of [
      [from, to],
      [to, from],
    ]) {
      const idx = halfEdges.length;
      halfEdges.push({
        from: s,
        to: e,
        angle: angleOf(sub(nodes[e], nodes[s])),
        visited: false,
      });
      if (!outgoing.has(s)) outgoing.set(s, []);
      outgoing.get(s)!.push(idx);
    }
  }

  // Ausgehende Kanten je Knoten nach Winkel sortieren -- Voraussetzung fuer
  // "naechste Kante im Uhrzeigersinn".
  for (const list of outgoing.values()) {
    list.sort((i, j) => halfEdges[i].angle - halfEdges[j].angle);
  }

  const twinOf = (idx: number) => (idx % 2 === 0 ? idx + 1 : idx - 1);

  const nextEdge = (idx: number) => {
    const he = halfEdges[idx];
    const list = outgoing.get(he.to) || [];
    if (!list.length) return -1;
    const backAngle = angleOf(sub(nodes[he.from], nodes[he.to]));
    // Die Kante, die im Winkel direkt vor der Rueckrichtung liegt.
    let best = -1;
    let bestDelta = Infinity;
    for (const cand of list) {
      let delta = backAngle - halfEdges[cand].angle;
      while (delta <= EPS) delta += Math.PI * 2;
      if (delta < bestDelta) {
        bestDelta = delta;
        best = cand;
      }
    }
    // Sackgasse: nur die Zwillingskante vorhanden -> zurueck.
    return best === -1 ? twinOf(idx) : best;
  };

  const rooms: Room[] = [];
  for (let start = 0; start < halfEdges.length; start++) {
    if (halfEdges[start].visited) continue;
    const cycle: number[] = [];
    let cur = start;
    let guard = 0;
    while (!halfEdges[cur].visited && guard++ < halfEdges.length + 2) {
      halfEdges[cur].visited = true;
      cycle.push(cur);
      cur = nextEdge(cur);
      if (cur < 0) break;
    }
    if (cycle.length < 3) continue;
    const points = cycle.map((i) => ({ ...nodes[halfEdges[i].from] }));
    const area = polygonArea(points);
    // Nur Flaechen mit positivem Umlaufsinn sind Innenraeume; die Aussenhuelle
    // laeuft andersherum und wird hier verworfen.
    if (area > 100) rooms.push({ points, area });
  }
  return rooms;
}
